src/python/pants/engine/internals/parser.py

```diff
@@ -64,19 +65,39 @@ class BuildFilePreludeSymbols(BuildFileSymbolsInfo):
     @classmethod
     def create(cls, ns: Mapping[str, Any], env_vars: Iterable[str]) -> BuildFilePreludeSymbols:
         info = {}
+        annotations = ns.get("__annotations__", {})
         for name, symb in ns.items():
-            info[name] = BuildFileSymbolInfo(name, symb)
+            if name.startswith("_"):
+                continue
+            # We only need type hints via `annotations` for top-level values which doesn't work with `inspect`.
+            info[name] = BuildFileSymbolInfo(name, symb, type_hints=annotations.get(name))
         return cls(info=FrozenDict(info), referenced_env_vars=tuple(sorted(env_vars)))
 
 
 @dataclass(frozen=True)
 class BuildFileSymbolInfo:
     name: str
     value: Any
+    help: str | None = field(default=None, compare=False)
 
-    @property
-    def help(self) -> str | None:
-        return inspect.getdoc(self.value) if hasattr(self.value, "__name__") else None
+    type_hints: InitVar[Any] = None
+
+    def __post_init__(self, type_hints: Any) -> None:
+        if self.help is not None:
+            return
+
+        if hasattr(self.value, "__name__"):
+            help = inspect.getdoc(self.value)
+
+        elif type_hints is not None:
+            if typing.get_origin(type_hints) is Annotated:
+                typ, doc = typing.get_args(type_hints)
+                help = f"[type: {typ.__name__}] {doc}"
```

Comment:
An alternative to formatting the type into the string might be to put it in `signature` somehow, e.g. either dynamically (as below) or passing the type hint through:

```python
    @property
    def signature(self) -> str | None:
        if not callable(self.value):
            return f": {type(self.value).__name__}"
        ...
```

This would then come out as `: Version` for `PANTS_VERSION` (I'm thinking the leading `:` could be an indicator of "this is a constant").

The `help` output and the online renderer (https://github.com/pantsbuild/pantsbuild.org/blob/67038eb6956fe4c8fc956884897efebac6e8fd8d/src/components/reference/BuildFileSymbol.jsx#L6-L7) could then be adapted to format this "natively"